PYIC-3510: Correlate data for F2F VC

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Correlate data for F2F VC

### Why did it change

The f2f VC is collected async. We weren't doing the usual checks on the data in contains to be sure it correlates with the data in the other VCs. This meant data was being sent to SPoT that was being rejected.

This change will do the correlation in the check-existing-identity lambda and send the user to the F2F fail page if the data doesn't match. The user will then have the opportunity to reset their identity.

We don't need to take any action of the data doesn't correlate if we're not on an F2F journey. This is an expected scenario and will be picked up by them not matching a profile and be sent to the reset-journey lambda separately.

This also moves the call to updateVcStatues higher in the handler so it's available for the correlation check. This way the F2F VC is included.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-3510](https://govukverify.atlassian.net/browse/PYIC-3510)


[PYIC-3510]: https://govukverify.atlassian.net/browse/PYIC-3510?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-3510: Correlate data for F2F VC

The f2f VC is collected async. We weren't doing the usual checks on the
data in contains to be sure it correlates with the data in the other
VCs. This meant data was being sent to SPoT that was being rejected.

This change will do the correlation in the check-existing-identity
lambda and send the user to the F2F fail page if the data doesn't match.
The user will then have the opportunity to reset their identity.

We don't need to take any action of the data doesn't correlate if we're
not on an F2F journey. This is an expected scenario and will be picked
up by them not matching a profile and be sent to the reset-journey
lambda separately.

This also moves the call to updateVcStatues higher in the handler so
it's available for the correlation check. This way the F2F VC is
included.

therealslimsladey commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -357,4 +369,25 @@ private List<String> extractTxnIdsFromCredentials(List<SignedJWT> credentials)
+            throws NoVcStatusForIssuerException, HttpResponseExceptionWithErrorBody {
+        if (!userIdentityService.checkNameAndFamilyNameCorrelationInCredentials(
+                userId, currentVcStatuses)) {
+            LOGGER.error("Failed to correlate names");
```
Structured messages with StringMapMessage are preferred?

therealslimsladey replied:
Discussed and discounted

therealslimsladey commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -357,4 +369,25 @@ private List<String> extractTxnIdsFromCredentials(List<SignedJWT> credentials)
+
+        if (!userIdentityService.checkBirthDateCorrelationInCredentials(
+                userId, currentVcStatuses)) {
+            LOGGER.error("Failed to correlate birth dates");
```
Structured messages with StringMapMessage are preferred?

Wynndow replied:
I'm not sure they actually add much in this scenario though, other than making the data harder to see. Makes sense if you're logging discreet data about something you might want to search on. Like "how many log messages where ci-score >= 5". But this is just info.

therealslimsladey reviewed (changes requested):
minor items

therealslimsladey commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -259,9 +261,17 @@ public Map<String, Object> handleRequest(JourneyRequest event, Context context)
                             HttpStatus.SC_INTERNAL_SERVER_ERROR,
                             ErrorResponse.FAILED_TO_SEND_AUDIT_EVENT)
                     .toObjectMap();
+        } catch (NoVcStatusForIssuerException e) {
```
Test?

Wynndow replied:
Whoops - have added. Good catch.

therealslimsladey commented on lambdas/check-existing-identity/src/main/java/uk/gov/di/ipv/core/checkexistingidentity/CheckExistingIdentityHandler.java:
```diff
@@ -357,4 +369,25 @@ private List<String> extractTxnIdsFromCredentials(List<SignedJWT> credentials)
+            return false;
+        }
+
+        if (!userIdentityService.checkBirthDateCorrelationInCredentials(
```
Test?

Wynndow replied:
Oh snap. I have `shouldReturnFailResponseForFaceToFaceIfBirthDatesDoNotCorrelate` but it's actually not testing it at all. Will fix up.

Wynndow committed:
PYIC-3510: Add test for no VC status

I missed this test case. Whoops.

Wynndow committed:
PYIC-3510: Use string map message for error log

This improves the logging when we can't correlate the data. It will
report on all the data that couldn't be correlated (names or date of
birth) and log it in a structured way for later analysis.

Wynndow committed:
PYIC-3510: Fix test

The test was mocking the names correlation method twice. My bad.

therealslimsladey reviewed (approved):
lgtm
